core/commonMain/src/kotlinx/serialization/Serializers.kt

```diff
@@ -35,48 +35,64 @@ public inline fun <reified T> SerializersModule.serializer(): KSerializer<T> {
 /**
  * Creates a serializer for the given [type].
  * [type] argument can be obtained with experimental [typeOf] method.
+ *  @throws SerializationException if serializer cannot be created (provided [type] or its type argument is not serializable).
  */
 @OptIn(ExperimentalSerializationApi::class)
-public fun serializer(type: KType): KSerializer<Any?> {
-    val result = EmptySerializersModule.serializerByKTypeImpl(type) ?: type.kclass().platformSpecificSerializerNotRegistered()
-    return result.nullable(type.isMarkedNullable)
-}
+public fun serializer(type: KType): KSerializer<Any?> = EmptySerializersModule.serializer(type)
+
+/**
+ * Creates a serializer for the given [type].
+ * [type] argument can be obtained with experimental [typeOf] method.
+ * Returns `null` if serializer cannot be created (provided [type] or its type argument is not serializable).
+ */
+@OptIn(ExperimentalSerializationApi::class)
+public fun serializerOrNull(type: KType): KSerializer<Any?>? = EmptySerializersModule.serializerOrNull(type)
 
 /**
  * Attempts to create a serializer for the given [type] and fallbacks to [contextual][SerializersModule.getContextual]
  * lookup for non-serializable types.
  * [type] argument can be obtained with experimental [typeOf] method.
+ * @throws SerializationException if serializer cannot be created (provided [type] or its type argument is not serializable and is not registered in [this] module).
  */
 @OptIn(ExperimentalSerializationApi::class)
-public fun SerializersModule.serializer(type: KType): KSerializer<Any?> {
-    val kclass = type.kclass()
-    val isNullable = type.isMarkedNullable
-    val builtin = serializerByKTypeImpl(type)
-    if (builtin != null) {
-        return builtin.nullable(isNullable).cast()
-    }
+public fun SerializersModule.serializer(type: KType): KSerializer<Any?> =
+    serializerByKTypeImpl(type, checkTypeArgs = true) ?: type.kclass().platformSpecificSerializerNotRegistered()
```

Comment:
Here and everywhere: `checkTypeArgs` tells us how exactly it works internally, not its intent, so it is required to study the implementation in order to understand why is it here.
I suggest to rename it to better reflect the intent e.g. `nullableSerializer`/`nullWhenMissing`/`failOnMissingSerializer`